frame/babe/src/lib.rs

```diff
@@ -331,19 +331,61 @@ pub mod pallet {
 	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
 		/// Initialization
 		fn on_initialize(now: BlockNumberFor<T>) -> Weight {
-			Self::do_initialize(now);
+			Self::initialize(now);
 			0
 		}
 
 		/// Block finalization
-		fn on_finalize(_n: BlockNumberFor<T>) {
+		fn on_finalize(_now: BlockNumberFor<T>) {
 			// at the end of the block, we can safely include the new VRF output
 			// from this block into the under-construction randomness. If we've determined
 			// that this block was the first in a new epoch, the changeover logic has
 			// already occurred at this point, so the under-construction randomness
 			// will only contain outputs from the right epoch.
-			if let Some(Some(randomness)) = Initialized::<T>::take() {
-				Self::deposit_randomness(&randomness);
+			if let Some(Some(pre_digest)) = Initialized::<T>::take() {
```

Comment:
```suggestion
			let pre_digest = Initialized::<T>::take().flatten().expect("BABE pre-digest is required to be present for valid BABE blocks");
```

The message is a little bit shitty and could  be improved, however when we reach this there should be a pre digest?

Reply:
Currently this is not guaranteed, there are a bunch of tests / benchmarks that create BABE blocks without pre-digest. I will change this and make this fail in `on_initialize` if there is no pre-digest, but would prefer to do that in follow-up PR.

Reply:
Sounds good. But you mean panic in on_finalize?

Reply:
(I would still keep the invariant that we don't fail on missing pre digests in on initialize)

Reply:
> (I would still keep the invariant that we don't fail on missing pre digests in on initialize)

I was thinking of panicking in `on_initialize` if the pre-digest is not found, this way we can avoid handling these `Option`s here. Or do you mean that you prefer to fail in `on_finalize` rather than `on_initialize`?

Reply:
> Or do you mean that you prefer to fail in `on_finalize` rather than `on_initialize`

Yes exactly.

I mean we currently don't call `on_initialize` anymore for each runtime call, but I would here be better on the safe side :D And as long as we panic when the pre-runtime digest is missing, we can still do it at the end of the block.

Reply:
That makes sense to me, I wasn't thinking of runtime calls.